In [1]:
import pandas as pd
from sklearn import metrics
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error
from itertools import combinations

In [2]:
housing = pd.read_csv('C:/Users/tejas/Downloads/housing_processed.csv')
housing.head()

Id  MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
0   1          60         65.0     8450            7            5       2003   
1   2          20         80.0     9600            6            8       1976   
2   3          60         68.0    11250            7            5       2001   
3   4          70         60.0     9550            7            5       1915   
4   5          60         84.0    14260            8            5       2000   

   YearRemodAdd  MasVnrArea ExterQual  ... GarageType_NA SaleType_COD  \
0          2003       196.0        Gd  ...             0            0   
1          1976         0.0        TA  ...             0            0   
2          2002       162.0        Gd  ...             0            0   
3          1970         0.0        TA  ...             0            0   
4          2000       350.0        Gd  ...             0            0   

  SaleType_CWD SaleType_Con SaleType_ConLD  SaleType_ConLI SaleType_ConLw  \
0            0            0              0               0              0   
1            0            0              0               0              0   
2            0            0              0               0              0   
3            0            0              0               0              0   
4            0            0              0               0              0   

   SaleType_New  SaleType_Oth  SaleType_WD  
0             0             0            1  
1             0             0            1  
2             0             0            1  
3             0             0            1  
4             0             0            1  

[5 rows x 238 columns]

In [3]:
housing[["ExterQual","ExterQual_Coded"]].head()

ExterQual  ExterQual_Coded
0        Gd                3
1        TA                2
2        Gd                3
3        TA                2
4        Gd                3

In [4]:
type(housing.dtypes==object)

pandas.core.series.Series

In [5]:
housing.dtypes[housing.dtypes==object]

ExterQual       object
ExterCond       object
BsmtQual        object
BsmtCond        object
BsmtExposure    object
BsmtFinType1    object
BsmtFinType2    object
HeatingQC       object
CentralAir      object
KitchenQual     object
Functional      object
FireplaceQu     object
GarageFinish    object
GarageQual      object
GarageCond      object
PavedDrive      object
dtype: object

In [6]:
housing.dtypes[housing.dtypes==object].shape

(16,)

In [7]:
"SalePrice" in housing.columns

True

In [8]:
len(housing.columns)

238

In [9]:
# We could drop columns by name:
housing_ml = housing.drop(columns=["ExterQual"])

In [10]:
housing_ml = housing.loc[:,housing.dtypes != object]

In [11]:
len(housing_ml.columns)

222

In [12]:
housing_ml.shape

(1460, 222)

In [13]:
# Split data as features and target
# take "SalePrice" values into its own 1-D array 
sale_price = housing_ml.pop('SalePrice')
type(sale_price)

pandas.core.series.Series

In [14]:
sale_price.shape

(1460,)

In [15]:
# pop removes the column
# "in place" operation
# now housing_ml has one less column
housing_ml.shape

(1460, 221)

In [16]:
y = sale_price.values
type(y)

numpy.ndarray

In [17]:
y.shape
# (1460,)
# it is equivalent to (1460)
# means it is a 1-d array

(1460,)

In [18]:

X = housing_ml.values

In [19]:
X.shape

(1460, 221)

In [20]:
X.shape[0] == y.shape[0]

True

In [21]:
# Import estimator
from sklearn.dummy import DummyRegressor
# Instantiate estimator
# guess the mean every single time
mean_reg = DummyRegressor(strategy='mean')
# fit estimator
mean_reg.fit(X, y)

DummyRegressor()

In [22]:
# predict
mean_reg.predict(X)

array([180921.19589041, 180921.19589041, 180921.19589041, ...,
       180921.19589041, 180921.19589041, 180921.19589041])

In [23]:
mean_reg.score(X, y)

0.0

In [24]:
from sklearn.linear_model import LinearRegression
linear_model = LinearRegression()

In [25]:
X_lf = housing_ml[['LotFrontage']]
X_lf

LotFrontage
0            65.0
1            80.0
2            68.0
3            60.0
4            84.0
...           ...
1455         62.0
1456         85.0
1457         66.0
1458         68.0
1459         75.0

[1460 rows x 1 columns]

In [26]:
linear_model.fit(X_lf, y)

LinearRegression()

In [27]:
linear_model.fit(X, y)
linear_model.score(X, y)

0.918472889091411

In [ ]:
y_pred = linear_model.predict(X_lf)
y_pred

In [30]:
X_lf.shape
linear_model.score(X_lf, y)

(1460, 1)

In [32]:
y

array([208500, 181500, 223500, ..., 266500, 142125, 147500], dtype=int64)

In [33]:
import matplotlib.pyplot as plt
%matplotlib inline
plt.figure(figsize=(12, 5))
plt.scatter(y, y_pred);

In [35]:
housing_ml

Id  MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  \
0        1          60         65.0     8450            7            5   
1        2          20         80.0     9600            6            8   
2        3          60         68.0    11250            7            5   
3        4          70         60.0     9550            7            5   
4        5          60         84.0    14260            8            5   
...    ...         ...          ...      ...          ...          ...   
1455  1456          60         62.0     7917            6            5   
1456  1457          20         85.0    13175            6            6   
1457  1458          70         66.0     9042            7            9   
1458  1459          20         68.0     9717            5            6   
1459  1460          20         75.0     9937            5            6   

      YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  ...  GarageType_NA  \
0          2003          2003       196.0         706  ...              0   
1          1976          1976         0.0         978  ...              0   
2          2001          2002       162.0         486  ...              0   
3          1915          1970         0.0         216  ...              0   
4          2000          2000       350.0         655  ...              0   
...         ...           ...         ...         ...  ...            ...   
1455       1999          2000         0.0           0  ...              0   
1456       1978          1988       119.0         790  ...              0   
1457       1941          2006         0.0         275  ...              0   
1458       1950          1996         0.0          49  ...              0   
1459       1965          1965         0.0         830  ...              0   

      SaleType_COD  SaleType_CWD  SaleType_Con  SaleType_ConLD  \
0                0             0             0               0   
1                0             0             0               0   
2                0             0             0               0   
3                0             0             0               0   
4                0             0             0               0   
...            ...           ...           ...             ...   
1455             0             0             0               0   
1456             0             0             0               0   
1457             0             0             0               0   
1458             0             0             0               0   
1459             0             0             0               0   

      SaleType_ConLI  SaleType_ConLw  SaleType_New  SaleType_Oth  SaleType_WD  
0                  0               0             0             0            1  
1                  0               0             0             0            1  
2                  0               0             0             0            1  
3                  0               0             0             0            1  
4                  0               0             0             0            1  
...              ...             ...           ...           ...          ...  
1455               0               0             0             0            1  
1456               0               0             0             0            1  
1457               0               0             0             0            1  
1458               0               0             0             0            1  
1459               0               0             0             0            1  

[1460 rows x 221 columns]

In [36]:
X=housing_ml

In [40]:
y=sale_price.values

In [41]:
#Splitting the data set into train and test
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.8, random_state=21)

In [42]:
X_train[['Id']].shape

(1168, 1)

In [43]:
X_train.dtypes

Id                  int64
MSSubClass          int64
LotFrontage       float64
LotArea             int64
OverallQual         int64
                   ...   
SaleType_ConLI      int64
SaleType_ConLw      int64
SaleType_New        int64
SaleType_Oth        int64
SaleType_WD         int64
Length: 221, dtype: object

In [44]:
#Training the model with each feauture one at a time
empty_df = pd.DataFrame(columns = ['Feauture', 'train_score','test_score' ,'RMSE'])
for i in X_train:
    column_name=i
    X_all=X_train[[i]]
    linear_model.fit(X_all, y_train)
    y_pred_all = linear_model.predict(X_test[[i]])
    feauture_scores=linear_model.score(X_all, y_train)
    test_feauture_scores=linear_model.score(X_test[[i]], y_test)
    Rmse=np.sqrt(metrics.mean_squared_error(y_test,y_pred_all))
    empty_df = empty_df.append({'Feauture' : column_name, 'train_score' : feauture_scores,'test_score': test_feauture_scores,'RMSE' : Rmse}, 
                ignore_index = True)

In [45]:
empty_df

Feauture  train_score  test_score          RMSE
0                Id     0.000750   -0.000807  82357.650982
1        MSSubClass     0.006301    0.009706  81923.935781
2       LotFrontage     0.111296    0.114481  77468.943523
3           LotArea     0.065862    0.079231  78995.796162
4       OverallQual     0.620145    0.645463  49018.435788
..              ...          ...         ...           ...
216  SaleType_ConLI     0.002522   -0.017066  83023.915840
217  SaleType_ConLw     0.001583   -0.003102  82452.021869
218    SaleType_New     0.127505    0.128701  76844.414403
219    SaleType_Oth     0.000642    0.002900  82204.964657
220     SaleType_WD     0.065960    0.031580  81014.083229

[221 rows x 4 columns]

In [46]:
#Sorting the top 10 test scores
final_df = empty_df.sort_values(by=['test_score'], ascending=False)
final_dff=final_df.head(10)
final_dff

Feauture  train_score  test_score          RMSE
4          OverallQual     0.620145    0.645463  49018.435788
36     ExterQual_Coded     0.456558    0.499175  58260.148655
16           GrLivArea     0.517889    0.444290  61369.555159
44   KitchenQual_Coded     0.433710    0.439703  61622.301138
12         TotalBsmtSF     0.363021    0.424913  62430.331676
13            1stFlrSF     0.353373    0.416668  62876.256239
25          GarageCars     0.409449    0.411931  63131.022101
26          GarageArea     0.383182    0.407326  63377.710181
38      BsmtQual_Coded     0.342356    0.340970  66831.518131
47  GarageFinish_Coded     0.297428    0.316285  68071.654684

In [47]:
#getting all combinations of 10 feautures
my_combi=list(combinations(final_dff.Feauture.values, 2))
my_combi[0]

('OverallQual', 'ExterQual_Coded')

In [48]:
#3Training the model with two features at a time
empty_df_final = pd.DataFrame(columns = ['Feauture1','Feauture2','train_score','test_score', 'RMSE'])
for z in my_combi:
    a=z[0]
    b=z[1]
    column_name1=a
    column_name2=b
    X_two=X_train[[a,b]]
    linear_model.fit(X_two, y_train)
    y_pred_two = linear_model.predict(X_test[[a,b]])
    feauture_scores_two=linear_model.score(X_two, y_train)
    test_feauture_scores_two=linear_model.score(X_test[[a,b]], y_test)
    Rmse_two=np.sqrt(metrics.mean_squared_error(y_test,y_pred_two))
    empty_df_final = empty_df_final.append({'Feauture1' : column_name1,'Feauture2' : column_name2, 'train_score' : feauture_scores_two, 'test_score': test_feauture_scores_two,'RMSE' : Rmse_two}, 
    ignore_index = True)

In [281]:
empty_df_final

Feauture1          Feauture2  train_score  test_score  \
0         OverallQual          GrLivArea     0.719790    0.692199   
1         OverallQual    ExterQual_Coded     0.645496    0.668015   
2         OverallQual  KitchenQual_Coded     0.650178    0.673087   
3         OverallQual         GarageCars     0.667212    0.672325   
4         OverallQual         GarageArea     0.669904    0.681395   
5         OverallQual        TotalBsmtSF     0.666766    0.706791   
6         OverallQual           1stFlrSF     0.686998    0.717388   
7         OverallQual     BsmtQual_Coded     0.633166    0.656666   
8         OverallQual           FullBath     0.646203    0.654407   
9           GrLivArea    ExterQual_Coded     0.684981    0.634781   
10          GrLivArea  KitchenQual_Coded     0.675586    0.606902   
11          GrLivArea         GarageCars     0.631802    0.596687   
12          GrLivArea         GarageArea     0.614104    0.596072   
13          GrLivArea        TotalBsmtSF     0.619065    0.571193   
14          GrLivArea           1stFlrSF     0.574026    0.524892   
15          GrLivArea     BsmtQual_Coded     0.657906    0.581269   
16          GrLivArea           FullBath     0.540654    0.461779   
17    ExterQual_Coded  KitchenQual_Coded     0.515648    0.561480   
18    ExterQual_Coded         GarageCars     0.569290    0.595875   
19    ExterQual_Coded         GarageArea     0.566341    0.599368   
20    ExterQual_Coded        TotalBsmtSF     0.562577    0.623021   
21    ExterQual_Coded           1stFlrSF     0.588295    0.636405   
22    ExterQual_Coded     BsmtQual_Coded     0.518269    0.545053   
23    ExterQual_Coded           FullBath     0.530668    0.550411   
24  KitchenQual_Coded         GarageCars     0.557762    0.567834   
25  KitchenQual_Coded         GarageArea     0.552700    0.557445   
26  KitchenQual_Coded        TotalBsmtSF     0.558783    0.599046   
27  KitchenQual_Coded           1stFlrSF     0.574853    0.596879   
28  KitchenQual_Coded     BsmtQual_Coded     0.514010    0.532859   
29  KitchenQual_Coded           FullBath     0.531333    0.513459   
30         GarageCars         GarageArea     0.423134    0.432905   
31         GarageCars        TotalBsmtSF     0.546273    0.557316   
32         GarageCars           1stFlrSF     0.535454    0.559115   
33         GarageCars     BsmtQual_Coded     0.521289    0.519033   
34         GarageCars           FullBath     0.504028    0.469618   
35         GarageArea        TotalBsmtSF     0.508228    0.538125   
36         GarageArea           1stFlrSF     0.497785    0.542283   
37         GarageArea     BsmtQual_Coded     0.519698    0.522865   
38         GarageArea           FullBath     0.505722    0.487358   
39        TotalBsmtSF           1stFlrSF     0.396284    0.454213   
40        TotalBsmtSF     BsmtQual_Coded     0.448427    0.501217   
41        TotalBsmtSF           FullBath     0.524662    0.515206   
42           1stFlrSF     BsmtQual_Coded     0.541718    0.573221   
43           1stFlrSF           FullBath     0.493288    0.497710   
44     BsmtQual_Coded           FullBath     0.485187    0.455962   

            RMSE  
0   45673.467009  
1   47433.824631  
2   47070.072304  
3   47124.910052  
4   46468.134901  
5   44577.664096  
6   43764.688318  
7   48237.754631  
8   48396.171085  
9   49751.391459  
10  51615.352295  
11  52281.742673  
12  52321.567243  
13  53908.788204  
14  56744.643168  
15  53271.697641  
16  60396.129092  
17  54515.960376  
18  52334.309055  
19  52107.664319  
20  50546.047839  
21  49640.643171  
22  55527.623429  
23  55199.694643  
24  54119.562563  
25  54766.149611  
26  52128.621701  
27  52269.257276  
28  56266.837646  
29  57423.326350  
30  61995.006197  
31  54774.150984  
32  54662.769998  
33  57093.457361  
34  59954.692652  
35  55948.807553  
36  55696.442890  
37  56865.592068  
38  58943.485438  
39  60819.169854  
40  58141.304193  
41  57320.176422  
42  53781.197519  
4

In [49]:
#sorting and saving the top 10 feautures
two_final_df = empty_df_final.sort_values(by=['test_score'], ascending=False)
two_final_dff=two_final_df.head(10)
two_final_dff

Feauture1           Feauture2  train_score  test_score          RMSE
4       OverallQual            1stFlrSF     0.686998    0.717388  43764.688318
3       OverallQual         TotalBsmtSF     0.666766    0.706791  44577.664096
1       OverallQual           GrLivArea     0.719790    0.692199  45673.467009
6       OverallQual          GarageArea     0.669904    0.681395  46468.134901
2       OverallQual   KitchenQual_Coded     0.650178    0.673087  47070.072304
5       OverallQual          GarageCars     0.667212    0.672325  47124.910052
0       OverallQual     ExterQual_Coded     0.645496    0.668015  47433.824631
7       OverallQual      BsmtQual_Coded     0.633166    0.656666  48237.754631
8       OverallQual  GarageFinish_Coded     0.640200    0.651902  48571.271663
12  ExterQual_Coded            1stFlrSF     0.588295    0.636405  49640.643171

In [50]:
4.#fitting the model with all feautures at a time
linear_model.fit(X_train, y_train)

LinearRegression()

In [51]:
#making predictions on test data
y_pred = linear_model.predict(X_test)

In [52]:
all_rmse=np.sqrt(mean_squared_error(y_test, y_pred))
all_rmse

35623.303450007385

In [53]:
#Score obtained with training data sets
linear_model.score(X_train,y_train)

0.9287445945544778

In [54]:
#Score obtained with test data sets
linear_model.score(X_test,y_test)

0.8127547098527904

In [55]:
#5 fittting the model with training set with K=5
from sklearn.neighbors import KNeighborsRegressor
neigh = KNeighborsRegressor(n_neighbors=5)
neigh.fit(X_train, y_train)

KNeighborsRegressor()

In [56]:
five_y_pred=neigh.predict(X_test)

In [57]:
five_all_rmse=np.sqrt(mean_squared_error(y_test,five_y_pred))
five_all_rmse

51966.07983368139

In [58]:
neigh.score(X_train, y_train)

0.7717466301028698

In [59]:
neigh.score(X_test, y_test)

0.6015421069563245

In [62]:
# fittting the model with training set with K=10
neigh10 = KNeighborsRegressor(n_neighbors=10)
neigh10.fit(X_train, y_train)

KNeighborsRegressor(n_neighbors=10)

In [64]:
ten_y_pred=neigh10.predict(X_test)

In [66]:
ten_all_rmse=np.sqrt(mean_squared_error(y_test,ten_y_pred))
ten_all_rmse

52392.16317256475

In [67]:
neigh10.score(X_train, y_train)

0.7130756140255491

In [68]:
neigh10.score(X_test, y_test)

0.5949812006956909

INFERENCE
1.Out of 5KNN and 10KNN regressors, 5KNN shows least RMSE and also has high accuracy than 10 KNN which may imply optimum K value lies between both

2.Linear regressor with one feauture seems better than N-KNN as it has low RMSE and high accuracy.

3.Linear Regressor with two features has better accuracy in prediction than one feature 

4.Overall When all the feautures are considered in the linear regressor the model shows high accuracy of 81% and also low error rate